plugins/module_utils/foreman_helper.py

```diff
@@ -643,10 +644,241 @@ def exit_json(self, changed=False, **kwargs):
+                if type(entity_dict[self.entity_scope]) is dict and 'id' in entity_dict[self.entity_scope]:
+                    scope_id = entity_dict[self.entity_scope]['id']
+                else:
+                    scope_id = entity_dict[self.entity_scope]
+                scope = "{0}_id = {1}".format(self.entity_scope, scope_id)
+                entity = self._resolve_entity(self.entity_name, value, entity_opts=self.entity_opts, params=scope)
+            else:
+                self.fail_json(msg='Scope {0} not found for {1}'.format(str(self.entity_scope), self.entity_name))
+        elif not self.entity_scope:
+            entity = self._resolve_entity(self.entity_name, value, entity_opts=self.entity_opts)
+        else:
+            self.fail_json(msg='Invalid scope {0} for {1}'.format(str(self.entity_scope), self.entity_name))
+
+        return entity
+
+    def _resolve_sub_entities(self, entity, entities, entity_dict):
+        for entity_name, entity_opts in {key: value for key, value in entities.items() if key in entity_dict}.items():
+            if 'scope' in entity_opts:
+                if entity_opts['scope'] == self.entity_name and entity:
+                    scope_id = entity['id']
+                elif entity_opts['scope'] in entity_dict:
+                    scope_id = entity_dict[entity_opts['scope']]['id']
+
+                scope = {'{}_id'.format(entity_opts['scope']): scope_id}
+            else:
+                scope = None
+            if entity_opts['type'] == 'entity' and type(entity_dict[entity_name]) is str:
+                entity_dict[entity_name] = self._resolve_entity(entity_name, entity_dict[entity_name], entity_opts, params=scope)
+            elif (entity_opts['type'] == 'entity_list' and type(entity_dict[entity_name]) is list
+                  and len(entity_dict[entity_name]) > 0 and type(entity_dict[entity_name][0]) is str):
+                entity_dict[entity_name] = self._resolve_entity_list(entity_name, entity_dict[entity_name], entity_opts, params=scope)
+
+    @property
+    def _unscoped_sub_entities(self):
+        return {key: value for key, value in self.sub_entities.items() if 'scope' not in value}
+
+    @property
+    def _scoped_sub_entities(self):
+        return {key: value for key, value in self.sub_entities.items() if 'scope' in value}
+
+    def _resolve_entity(self, entity_name, value, entity_opts, params=None):
+        p = apypie.Inflector()
+        resource_path = entity_opts.get('resource_type', p.pluralize(entity_name))
+        if resource_path == 'operatingsystems':
+            return self.find_operatingsystem(value, params=params,
+                                             failsafe=entity_opts.get('failsafe', False),
+                                             thin=entity_opts.get('thin', True))
+        else:
+            return self.find_resource_by(resource_path,
+                                         entity_opts.get('search_by', 'name'), value,
```

Comment:
IMHO we can set the `name` default in the `__init__`, see suggestion there.

Reply:
This is used to resolve sub entities too, and I didn't set defaults for all sub entities for `search_by` `search_operator`, `thin` and `failsafe`